src/globus_sdk/services/gcs/data/_common.py

```diff
@@ -7,16 +7,25 @@
     from typing import Protocol
 
 
+VersionTuple = t.Tuple[int, int, int]
+
+DatatypeCallback = t.Callable[["DocumentWithInducedDatatype"], t.Optional[VersionTuple]]
+
+
 class DocumentWithInducedDatatype(Protocol):
     DATATYPE_BASE: str
-    DATATYPE_VERSION_IMPLICATIONS: t.Dict[str, t.Tuple[int, int, int]]
+    DATATYPE_VERSION_IMPLICATIONS: t.Dict[str, VersionTuple]
+    DATATYPE_VERSION_CALLBACKS: t.Tuple[DatatypeCallback, ...]
 
     def __contains__(self, key: str) -> bool:  # pragma: no cover
         ...
 
     def __setitem__(self, key: str, value: t.Any) -> None:  # pragma: no cover
         ...
 
+    def __getitem__(self, key: str) -> t.Any:  # pragma: no cover
+        ...
+
```

Comment:
A brief note for reviewers:

I'm well aware that this protocol looks like it could inherit from Mapping. I've avoided making that change for two basic reasons:
- I don't fully know what the implications are of protocol inheritance (normal class inheritance rules?)
- Protocols are a way of getting non-inheritance based polymorphism. Using inheritance to define them strikes me instinctively as wrong, although maybe it's fine.

On both points, I'm mostly unsure and avoiding the issue for now.